keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,1053 @@
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+- **Restricted state transition:** The `schedulingGates` field can be initialized only when a Pod is
+created (either by the client, or mutated during admission). After creation, each `schedulingGate`
+can be removed in arbitrary order, but addition of a new scheduling gate is disallowed.
+To ensure consistency, scheduled Pod must always have empty `schedulingGates`.
+This means that a client (an administrator or custom controller) cannot create a Pod that has
+`schedulingGates` populated in any way, if that Pod also specified a `spec.nodeName`. In this case,
+API Server will return a 409 conflict / validation error.
+
+    |                                              | non-nil schedulingGates⏸️ | nil schedulingGates▶️ |
+    |----------------------------------------------|-------------------------|-------------------------|
+    | unscheduled Pod<br>(nil <tt>nodeName</tt>)   | ✅ create<br>❌ update   | ✅ create<br>✅ update    |
+    | scheduled Pod<br>(non-nil <tt>nodeName</tt>) | ❌ create<br>❌ update   | ✅ create<br>✅ update    |
+
+- **New field disabled in Alpha but not scheduler extension:** In Alpha, the new Pod field is disabled
+by default. However, the scheduler's extension point is activated no matter the feature gate is enabled
+or not. This enables scheduler plugin developers to tryout the feature even in Alpha, by crafting
+different enqueue plugins and wire with custom fields or conditions.
+
+- **New column in kubectl:** To provider better UX, we're going to add a column "scheduling paused"
+to the output of `kubectl get pod` to indicate whether it's scheduling-paused or not.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- The scheduler doesn't actively clear a Pod's `schedulingGates` field. This means if some controller
+is ill-implemented, some Pods may stay in Pending state incorrectly. If you noticed a Pod stays
+in Pending state for a long time and it carries non-nil `schedulingGates`, you may find out which
+component owns the gate(s) (via `.spec.managedFields`) and report the symptom to the component owner.
+
+- Faulty controllers may forget to remove the Pod's `schedulingGates`, and hence results in
+a large number of unschedulable Pods. In Alpha, we don't limit the number of unschedulabe Pods caused
+by potential faulty controllers. We will evaluate necessary options in the future to mitigate
+potential abuse.
```

Comment:
The feature itself doesn't cause creating net new pods, so I don't think we need an explicit mechanism to handle this risk. This is similar to the scheduler continuously failing to schedule pods because the cluster doesn't have enough resources (e.g., a faulty cluster autoscaler or a faulty storage provisioning controller).

I think a guiding example here is finalizers, IIRC we don't have a mechanism to deal with finalizers that don't get cleared for a while.

Reply:
> I think a guiding example here is finalizers, IIRC we don't have a mechanism to deal with finalizers that don't get cleared for a while.

This is yet another thing that (in case of bugs) can result in leaking resources.
I'm not saying that we necessarily have to introduce some automatic removal of feature gates. But we need to introduce some mechanism to at least let people know that this is happening, e.g. a metric that exposes number of "notReady" pods or sth like that, so that as an operator, I can at least keep track of it.

Reply:
Yeah, we have the metric from the scheduler and a default pod condition (@Huang-Wei I don't see that yet in the KEP, can we add it please).

Reply:
@ahg-g Yes, metric was added but I missed the default pod condition part. Updated now. PTAL.